datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+    """ Extend `cls` to supports the "kill-interface"
+
+    This function creates a subclass of `cls` that contains the components
+    of the "kill-interface".  The two major components are amethod called `arm`,
+    and logic inside the timeout handler that can trigger a termination or
+    kill signal to the subprocess, if the termination time or kill time has
+    come.
+
+    Parameters
+    ----------
+    cls : type[Protocol]
+        A protocol class that should be extended by the kill-interface
+
+    Returns
+    -------
+    Any
+        If the `cls` is not a subclass of `GeneratorMixIn`, the result of
+        `cls._prepare_result` will be returned.
+
+    Generator
+        If the protocol is a subclass of `GeneratorMixIn`, the `Generator` that
+        is a result of `ThreadedRunner.run()` will be returned.
+    """
+    class KillWrapper(cls):
+        def __init__(self, *args, **kwargs):
+            self.armed = kwargs.pop('armed')
+            self.introduced_timeout = kwargs.pop('introduced_timeout')
+            self.terminate_time = kwargs.pop('terminate_time')
+            kill_time = kwargs.pop('kill_time')
+            self.kill_time = (
+                ((self.terminate_time or 0) + kill_time)
+                if kill_time is not None
+                else kill_time
+            )
+            self.process: subprocess.Popen | None = None
+            self.kill_counter: int = 0
+            super().__init__(*args, **kwargs)
+
+        def arm(self) -> None:
+            self.kill_counter = 0
+            self.armed = True
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed:
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+                self.kill_counter += 1
+
+            # If we set the timeout argument due to a not None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+        cmd: list,
+        protocol_class: type[Protocol],
+        *,
+        cwd: Path | None = None,
+        input: int | IO | bytes | Queue[bytes | None] | None = None,
+        timeout: float | None = None,
+        terminate_time: int | None = None,
+        kill_time: int | None = None,
```

Comment:
I think we need a precise definition in the documentation.

What is the difference between `timeout` and `terminate_time`?

Is `kill_time` on top of `terminate_time`?

What is the semantic difference that leads to one being called `timeout` and the other just be `time`?

Reply:
I want to add a few more comments after having tried to exercise this through in a concrete use case.

It is unclear where the timeouts apply, and thereby it is unclear how the  must be used.

Example: I have a process that downloads some information. This information in reported in a json-lines format. The process is batch-mode of some kind (an input triggers a response, continued until stdin closes).

Now there are several layers a user of this helper needs to be aware of.

- the process that downloads and eventually reports a JSON-line
- the protocol that assembles a JSON-line and decodes it, once complete

Let's assume the remote bandwidth is really low (few bytes per second), and the information that needs to go into a single JSON-line is a few kb, so it needs minutes to assemble a response, and have the "outer" protocol generator yield something.

What would the timeouts apply to? The inner layer (i.e. nothing coming down the remote connection), or the outer layer (no JSON-line-based result yielded for X seconds)?

Reply:
If you use a generator-protocol, then the timeouts that aid the terminating and killing mechanism are only triggered when the context is left. I extended the documentation to clarify that. So for `stdout_batchcommand` and `annexjson_batchcommand`, the timeout-counting starts when the context is left, which means when the code in the context decides that all operations are done and exits the context.

Timeout-handling for the transmission should be done in the protocol that is used, e.g.  `StdOutCaptureGeneratorProtocol`. For example, the protocol might send timeout-messages to the result generator and the in-context code can decide what to do. In another example, if the protocol raises an exception in its `timeout`-callback, the `batchcommand`-context would be left and the terminating- and killing- timeouts would be started.

Reply:
I understand the first paragraph and appreciate the clarification.

I have difficulties following the reasoning in the 2nd paragraph. When you say "should be done", you are saying "it is not implemented anywhere"? Do we have any protocol implementation that behaves like this? If not, how do we know that it would work?